cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java

```diff
@@ -951,16 +954,20 @@ private void writeHistory(Metadata metadata, long time) {
         byte[] rowKey = row.getRow();
         MetadataEntity metadataEntity = MetadataKey.extractMetadataEntityFromKey(rowKey);
         String metadataKey = MetadataKey.extractMetadataKey(rowKey);
-        Set<Indexer> indexers = getIndexersForKey(metadataKey);
+        Set<Indexer> indexers;
+        if (visitedEntities.contains(metadataEntity)) {
```

Comment:
this can be simplified into:

```
boolean isNewEntity = visitedEntities.add(metadataEntity);
Set<Indexer> indexers = getIndexersForKey(metadataKey, isNewEntity);
```